extensions/amp-a4a/0.1/amp-ad-templates.js

```diff
@@ -0,0 +1,124 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {dev} from '../../../src/log';
+import {getMode} from '../../../src/mode';
+import {getAmpdoc} from '../../../src/service';
+import {Bind} from '../../amp-bind/0.1/bind-impl';
+
+/** @private {!Object<string, string|boolean>} */
+const TEMPLATE_CORS_CONFIG = {
+  mode: 'cors',
+  method: 'GET',
+  // This should be cached across publisher domains, so don't append
+  // __amp_source_origin to the URL.
+  ampCors: false,
+  credentials: 'omit',
+};
+
+/** @const {string} */
+const TAG = 'amp-ad-templates';
+
+export class AmpAdTemplates {
+
+  /**
+   * @param {!Window} win
+   * @param {function(string)=} opt_onRetrieve Function to be executed upon
+   *   the successful retrieval of a template. If one is not supplied, we will
+   *   default to a no-op.
+   */
+  constructor(win, opt_onRetrieve = unusedTemplate => {}) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {LRUCache} */
+    this.cache_ = new LRUCache(5);
+  }
+
+  /**
+   * Fetch and parse template from AMP cache.  Result is stored in global in
+   * order to reduce overhead when template is used multiple times.
+   * @param {string} templateUrl CDN Proxy URL to template.
+   * @return {!Promise<string>}
+   */
+  fetch(templateUrl) {
+    let templatePromise = this.cache_.get(templateUrl);
+    if (!templatePromise) {
+      templatePromise = Services.xhrFor(this.win_)
+          .fetchText(getMode(this.win_).localDev
+            ? `http://ads.localhost:${this.win_.location.port}` +
+                `/a4a_template/adzerk/${templateUrl}`
+            : templateUrl, TEMPLATE_CORS_CONFIG)
+          .then(response => response.text());
+      this.cache_.put(templateUrl, templatePromise);
+    }
+    dev().assert(templatePromise);
+    return /** @type{!Promise<string>} */ (templatePromise);
+  }
+
+  /**
+   * @param {!JsonObject} templateValues The values to macro in.
+   * @param {!Element} element Parent element containing template.
+   */
+  render(templateValues, element) {
+    const win = element.ownerDocument.defaultView;
+    if (win) {
+      const bind = new Bind(getAmpdoc(element), win);
+      bind.setState(templateValues);
+    }
+  }
+}
+
+/** @typedef {{
+      payload: *,
+      access: number,
+    }} */
+export let Cacheable;
+
+class LRUCache {
+  /** @param {number} capacity */
+  constructor(capacity) {
```

Comment:
you can pass in a `generator` function in constructor. Then we only need one public method for the LRUCache: get()

Also, worth to extract the code into `src/utils/cache.js` for sharing.

Reply:
I'm not really sure how a `generator` would fit in here. Can you give an example of how you imagine the cache being used? If it's not too strong a preference on your end, I'd prefer to keep it with get/put.

Reply:
```js
class LRUCache {
  constructor(capacity, generator);

  get(id) {
    ...

    if (!this.cache_[id]) {
      this.cache_[id] = generator(id);
    }
    return this.cache_[id];    
  }
}
```

something like this

Reply:
Sorry, I'm still not understanding. Who calls `next()`, and how does this eliminate the need for `put()`?

Reply:
Since this is holding up release. Let's change this in a future PR. I don't think anyone will be using this code for a while anyway.

Reply:
it will be used like:

```
constructor() {
  this.cache_ = new LRUCache(5, url => {
      return Services.xhrFor(this.win_).fetchText(url);
  });
}
...

fetch(url) {
  return this.cache_.get(url);
}
```

there are other more important comments not addressed in this PR, so realistically, i don't think it should hold the release.

Reply:
Forgot to mark some of the comments as done. Will address the others shortly.